### DISCUSSION TITLE
New Feature! Pellet Manager (BETA)

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#21] [Show and tell]
For anyone who is following along, in the developer branch, I've just added a very basic pellet manager feature.  

* Allows you to track what kind of pellets you currently have in your smoker/grill.  
* You can add / remove brands of pellets from your database. 
* You can add / remove types of wood from your database.
* You can add profiles with a Brand / Wood Type / Rating & Comments. 
* Added Pellet / Hopper Level indicator for future implementation of pellet level sensor.  
* Added Pellet Info / Link to the dashboard.  
* Pellet Log / History, so you can see when you loaded a particular pellet type to your smoker

![PiFire-Pellet-Manager-00](https://user-images.githubusercontent.com/19876886/104265227-f716b500-5441-11eb-885e-1bdc9d29f1dd.png)
![PiFire-PelletManager-01](https://user-images.githubusercontent.com/19876886/104265241-fbdb6900-5441-11eb-9773-01ed157a2ccd.png)
![PiFire-PelletManager-02](https://user-images.githubusercontent.com/19876886/104265250-ff6ef000-5441-11eb-90f7-aa21be2a2a46.png)

I hope you like it!  Please let me know if you run into issues with this.  It will likely hit the main branch in a week or two.

**Comment by faisalhusain007** [comment]
Looks really awesome, I don't think there is anything like this even commercially. A highly useful feature for sure. I will test it soon.

Thank you for dedicating your time and effort to this project and keeping it alive.

Some new features I would also like to see would be an extension to the timer to turn the smoker off. Not sure if that would go in the recipies area for different cook times at different stages based on the probe temps in the protein being cooked.

**Reply by nebhead** [reply]
Thank you!  Yeah, I saw a post of someone who had a physical device that indicated what pellets were in their hopper (a little 3D printed rotating wheel), and I thought it was pretty cool.  It struck me then, that I could incorporate it into the WebUI pretty easily.  So, it stemmed from there and I started to build it out.  In the future, it might even be possible to gauge how many pellets were used by calculating the total auger time.  That might be a way to determine pellet levels without having to have an active sensor for that! 

Anyway, I'm going to go back to working on the Smoke Plus and physical button input.  That's probably going to take some more time/effort due to the fact that it requires me to work on the physical device (vs. the web interface which I can simulate on my PC).

**Reply by nebhead** [reply]
By the way, I just realized I didn't respond to the feature request to turn the smoker off when the timer expires.  I can certainly implement something like this outside of the recipe feature.  Honestly, I was a bit daunted by attempting to implement a recipe feature so hadn't tackled that yet.  I will get to that someday, but in the meantime can definitely put a tick-box in the timer modal to turn off the grill when the timer expires.

**Reply by faisalhusain007** [reply]
That would be great. Really appreciate your time and effort. I have been playing with the TOF VL53L sensor, it looks it could work placing it at the top and measuring the distance from a square piece of card placed at the top of the pellets not small enough to get caught in the auger and not too large to get in the way of the flow of pellets.

**Reply by nebhead** [reply]
BTW - I implemented the timer shutdown feature last night, but haven't pushed it to GitHub yet due to the fact that I'm still heavily tweaking the control flow to alleviate the previous errors I spoke about.  I think I have it nailed down, but I need to test it on the RaspberryPi HW.  Will probably push the code later today if everything looks good.

**Reply by faisalhusain007** [reply]
Awesome can't wait to test it, I just had a thought, could it be possible to also link the same shutdown feature to the food temperature probes when they reach the set temperature ?

**Reply by nebhead** [reply]
Yep, definitely possible to do this per probe similar to how I did it for the timer.  I'll try to add that as well before I push.  I had another thought that I'd like to separate the grill setpoint temp from the notification temp so that I can set the notification to occur before the grill gets to a certain setpoint.
